erpnext/vehicles/report/vehicle_maintenance_reminder/vehicle_maintenance_reminder.py

```diff
@@ -0,0 +1,182 @@
+# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
+# For license information, please see license.txt
+
+from __future__ import unicode_literals
+import frappe
+from frappe import _
+from frappe.utils import getdate, today, flt, add_months, today
+from dateutil.relativedelta import relativedelta
+from datetime import datetime
+
+def execute(filters=None):
+	return VehicleMaintenanceReminder(filters).run()
+
+
+class VehicleMaintenanceReminder:
+	def __init__(self, filters=None):
+		self.filters = frappe._dict(filters or {})
+		self.filters.from_date = getdate(filters.from_date or today())
+		self.filters.to_date = getdate(filters.to_date or today())
+
+		if self.filters.from_date > self.filters.to_date:
+			frappe.throw(_("Date Range is incorrect"))
+
+	def run(self):
+		self.get_data()
+		self.process_data()
+		self.get_columns()
+		return self.columns, self.data
+
+	def get_data(self):
+		self.project_template_data = frappe.db.sql("""
+			SELECT name as project_template, project_template_name, project_template_category, applies_to_item, due_after
+			FROM `tabProject Template`
+			WHERE due_after > 0 and ifnull(applies_to_item, '') != ''
+			ORDER BY due_after ASC
+		""", as_dict=1)
+
+		vehicle_or_conditions = self.process_project_template_data()
+
+		vehicle_data = frappe.db.sql("""
+			SELECT
+				name as vehicle, item_code, item_name, customer, customer_name, delivery_date,
+				chassis_no, engine_no, license_plate, unregistered, variant_of, variant_of_nam e
+			FROM `tabVehicle`
+			WHERE {0} and ifnull(delivery_document_no, '') != ''
+		""".format(vehicle_or_conditions), self.filters, as_dict=1)
+
+		self.data = vehicle_data
+
+
+	def process_project_template_data(self):
+		vehicle_or_conditions = []
+		applies_to_item_cache = frappe._dict()
+
+		for d in self.project_template_data:
+			template_condition = []
+			if d.applies_to_item:
+				if d.applies_to_item not in applies_to_item_cache:
+					item_codes = frappe.db.get_all("Item", filters={'variant_of': d.applies_to_item}, fields="item_code")
+					item_codes = [d.applies_to_item] + [i.item_code for i in item_codes]
+					applies_to_item_cache[d.applies_to_item] = item_codes
```

Comment:
You can move the logic of getting `item_codes` from `applies_to_item` into another function to make it more readable. You can also put `applies_to_item_cache` in the object to make it persistent